doc/man7/EVP_KDF-ARGON2.pod

```diff
@@ -0,0 +1,192 @@
+
+These parameters work as described in L<EVP_KDF(3)/PARAMETERS>.
+
+Note that RFC 9106 recommends 128 bits salt for most applications, or 64 bits
+salt in the case of space constraints. At least 128 bits output length is
+recommended.
+
+Note that secret (or pepper) is an optional secret data used along the
+password.
+
+=item "threads" (B<OSSL_KDF_PARAM_THREADS>) <unsigned integer>
+
+The number of threads, bounded above by the number of lanes.
+
+This can only be used with built-in thread support. Threading must be
+explicitly enabled. See EXAMPLES section for more information.
+
+=item "ad" (B<OSSL_KDF_PARAM_ARGON2_AD>) <octet string>
+
+Optional associated data, may be used to "tag" a group of keys, or tie them
+to a particular public key, without having to modify salt.
+
+=item "lanes" (B<OSSL_KDF_PARAM_ARGON2_LANES>) <unsigned integer>
+
+Argon2 splits the requested memory size into lanes, each of which is designed
+to be processed in parallel. For example, on a system with p cores, it's
+recommended to use p lanes.
+
+The number of lanes is used to derive the key. It is possible to specify
+more lanes than the number of available computational threads. This is
+especially encouraged if multi-threading is disabled.
+
+=item "memcost" (B<OSSL_KDF_PARAM_ARGON2_MEMCOST>) <unsigned integer>
+
+Memory cost parameter (the number of 1k memory blocks used).
+
+=item "version" (B<OSSL_KDF_PARAM_ARGON2_VERSION>) <unsigned integer>
+
+Argon2 version. Supported values: 0x10, 0x13 (default).
+
+=item "early_clean" (B<OSSL_KDF_PARAM_EARLY_CLEAN>) <unsigned integer>
+
+If set (nonzero), password and secret stored in Argon2 context are zeroed
+early during initial hash computation, as soon as they are not needed.
+Otherwise, they are zeroed along the rest of Argon2 context data on clear,
+free, reset.
+
+This can be useful if, for example, multiple keys with different ad value
+are to be generated from a single password and secret.
+
+=back
+
+=head1 EXAMPLES
+
+This example uses Argon2d with password "1234567890", salt "saltsalt",
+using 2 lanes, 2 threads, and memory cost of 65536:
+
+ #include <string.h>                 /* strlen               */
+ #include <openssl/core_names.h>     /* OSSL_KDF_*           */
+ #include <openssl/params.h>         /* OSSL_PARAM_*         */
+ #include <openssl/thread.h>         /* OSSL_set_max_threads */
+ #include <openssl/kdf.h>            /* EVP_KDF_*            */
+
+ int main(void)
+ {
+     int retval = 1;
+
+     EVP_KDF *kdf = NULL;
+     EVP_KDF_CTX *kctx = NULL;
+     OSSL_PARAM params[6], *p = params;
+
+     /* argon2 params, please refer to RFC9106 for recommended defaults */
+     uint32_t lanes = 2, threads = 2, memcost = 65536;
+     char pwd[] = "1234567890", salt[] = "saltsalt";
+
+     /* derive result */
+     size_t outlen = 128;
+     unsigned char result[outlen];
+
+     /* required if threads > 1 */
+     if (OSSL_set_max_threads(NULL, threads) != 1)
```

Comment:
Sorry to jump in. I found this pull request because I traced down the internal API: ossl_crypto_thread_start()/_join(), which were merged into master branch some time ago. Actually I'm thinking about another use case of multi-threads. Then, it leads me to this question: how can you be sure Argon2 is the only user of threads? how can you be so sure to set the max number of threads only according to only Argon2's need?

What if there is other users (maybe internal ossl frameworks) who set the max number already? How to address such conflicts?

Reply:
Not having looked at any of the implementations, and the manpage doesn't say anything about it, but I would expect the OSSL_set_max_threads() call to set a limit per libctx. So you can create a different libctx for other usecases, and that call should not use NULL for the libctx.

Reply:
> the OSSL_set_max_threads() call to set a limit per libctx.

Right, it sets a limit per libctx. This means within one libctx, there can be only one 'usecase' of the threadpool. How could this be justified? Suppose one method from one provider uses threadpool, and another method from another provider uses threadpool too. Then, these two providers cannot be loaded into the same libctx. So, here we are setting _implicit_  rules about loading providers.

I don't think this is expected, isn't it?

Reply:
We have to assume that the user and creator of the libctx knows what it is going to be used for and what sensible limits should be enforced.

> Then, these two providers cannot be loaded into the same libctx.

Err...yes they can? They may be sharing a thread pool but that isn't a problem per se if the pool has been adequately sized.

Although that does make me think of a different problem. Many (most?) providers will create their own libctx independent of the user libctx (or a child libctx based on the user one). Those libctx's will have a separate thread pool - which is fine (and means different providers won't interfere with each other) but does mean then end user has no real control over the "total" size of the thread pool.

Anyway this is kind of tangential to this Argon2 PR so discussion should probably be in some separate issue.

Reply:
> They may be sharing a thread pool but that isn't a problem per se if the pool has been adequately sized.

Still, they can steal the thread (passing the per user's upper limit, not max) from each other, right? I put more details in a new issue: https://github.com/openssl/openssl/issues/20059

> Anyway this is kind of tangential to this Argon2 PR so discussion should probably be in some separate issue.

Agree, and I submitted a new issue https://github.com/openssl/openssl/issues/20059

Reply:
If that's fine with you, @docularxu, I am resolving this conversation and we can continue the conversation in the two other issues.